File: crop_generator/cropgenerator/generatorobjects/generatorobjects.py
# Class definition for objects used in the crop_generator module and database
# Author: Michael B. Lance
# created: April 4, 2025
# updated: October 28, 2025
#---------------------------------------------------------------------------------------------------------------------------#

import numpy as np
from flask.json.provider import JSONProvider
import cv2
from abc import ABC, abstractmethod
from datetime import datetime
import io
import PIL.Image as PillowImage
from dataclasses import dataclass
from uuid import UUID
from flask_login import UserMixin 
from typing import Optional 

#---------------------------------------------------------------------------------------------------------------------------#
# ABC for easy serialization of child classes

class CgOBJ(ABC):
	@abstractmethod
	def serialize(self) -> dict:
		pass
	
#---------------------------------------------------------------------------------------------------------------------------#
# Project Management -- For Database use only

@dataclass
class Project(CgOBJ):
	''' Dataclass for representing projects from the databse
	
	'''
	project_id: int
	name: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'project_id': self.project_id,
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}

@dataclass
class Schema(CgOBJ):
	''' Dataclass for representing scehmas from the databse
	
	'''
	schema_id: int
	name: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'schema_id': self.schema_id,
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}

@dataclass
class Label(CgOBJ):
	''' Dataclass for representing labels from the databse
	
	'''
	label_id: int
	schema_id: int
	label: int
	name: str
	image_link: str
	created: datetime
	modified: datetime
	color: str
	uuid: UUID

	def serialize(self):
		return {
			'label_id': self.label_id,
			'label': self.label,
			'name': self.name,
			'image_link': self.image_link,
			'color': self.color,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}

@dataclass
class HerdUnit(CgOBJ):
	''' Dataclass for representing Herd Units from the database
	
	'''
	herd_unit_id: int
	name: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'herd_unit_id': self.herd_unit_id,
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}

@dataclass
class Model(CgOBJ):
	''' Dataclass for representing Models from the database
	
	'''
	model_id: int
	schema_id: int
	name: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'model_id': self.model_id,
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}
	
@dataclass
class Survey(CgOBJ):
	''' Dataclass for representing Surveys from the database
	
	'''
	survey_id: int
	survey_date: datetime
	name: str
	additional_info: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'survey_id': self.survey_id,
			'survey_date': self.survey_date,
			'name': self.name,
			'additional_info': self.additional_info,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}
	
#---------------------------------------------------------------------------------------------------------------------------#
# User Management -- For Database use only
	
@dataclass
class Role(CgOBJ):
	role_id: int
	name: str
	created: datetime
	modified: datetime
	uuid: UUID

	def serialize(self):
		return {
			'role_id': self.role_id,
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}
	
@dataclass
class Organization(CgOBJ):
	organization_id: int
	name: str
	created: datetime
	modified: datetime
	logo_url: str | None
	uuid: UUID

	def serialize(self):
		return {
			'organization_id': self.organization_id,
			'name': self.name,
			'created': self.created,
			'modified': self.modified,
			'logo_url': self.logo_url,
			'uuid': self.uuid
		}

class User(UserMixin, CgOBJ):
	def __init__(self, user_id: int, username: str, external_auth_id: str, external_auth_provider: str, status: str,
				created: datetime, modified: datetime, last_login: datetime,  locale: str, uuid: UUID, roles: Optional[list[Role]]=None):
		self.id = str(uuid) # this is this way to make Flask-Login happy
		self.user_id = user_id
		self.username = username
		self.external_auth_id = external_auth_id
		self.external_auth_provider = external_auth_provider
		self.status = status
		self.created = created
		self.modified = modified
		self.last_login = last_login
		self.locale = locale
		self.uuid = uuid
		self.roles = roles
	
	def getId(self) -> str:
		return self.id
	
	def hasRole(self, role_name: str) -> bool:
		if self.roles:
			return role_name in self.roles
		else:
			return False

	def serialize(self):
		return {
			'user_id': self.user_id,
			'username': self.username,
			'status': self.status,
			'created': self.created,
			'modified': self.modified,
			'last_login': self.last_login,
			'locale': self.locale,
			'uuid': self.id,
			'roles': self.roles
		}

#---------------------------------------------------------------------------------------------------------------------------#
# Core 

class Box(CgOBJ):
	''' A Box contains dimensional data for crops and predictions
	
	'''
	def __init__(self, top_left: tuple[int, int], bottom_right: tuple[int, int]):
		self.top_left = top_left
		self.bottom_right = bottom_right

	def getCenter(self) -> tuple:
		x = np.mean([self.top_left[0], self.bottom_right[0]])
		y = np.mean([self.top_left[1], self.bottom_right[1]])

		return ((abs(x), abs(y)))

	def getPoints(self) -> list[int]:
		return [self.top_left[0], self.top_left[1], self.bottom_right[0], self.bottom_right[1]]

	def calcIou(self, box_2) -> float:
		# Slightly modified from https://machinelearningspace.com/intersection-over-union-iou-a-comprehensive-guide/
		#Extract bounding boxes coordinates
		x0_A, y0_A, x1_A, y1_A = self.getPoints()
		x0_B, y0_B, x1_B, y1_B = box_2.getPoints()
		
		# Get the coordinates of the intersection rectangle
		x0_I = max(x0_A, x0_B)
		y0_I = max(y0_A, y0_B)
		x1_I = min(x1_A, x1_B)
		y1_I = min(y1_A, y1_B)
		#Calculate width and height of the intersection area.
		width_I = x1_I - x0_I 
		height_I = y1_I - y0_I

		# Handle the negative value width or height of the intersection area
		width_I = 0 if width_I < 0 else width_I
		height_I = 0 if height_I < 0 else height_I
		# Calculate the intersection area:
		intersection = width_I * height_I
		# Calculate the union area:
		width_A, height_A = x1_A - x0_A, y1_A - y0_A
		width_B, height_B = x1_B - x0_B, y1_B - y0_B
		union = (width_A * height_A) + (width_B * height_B) - intersection
		# Calculate the IoU:
		return intersection/union
	
	def serialize(self) -> dict:
		return {
			'top_left': {'x': self.top_left[0], 'y': self.top_left[1]},
			'bottom_right': {'x': self.bottom_right[0], 'y': self.bottom_right[1]},
		}

@dataclass
class Image(CgOBJ):
	def __init__(self, image_id: int, herd_unit_id: int, survey_id: int, name: str, img_key: str,
				in_training: bool, crops_generated: int, opened_by_user_id: int,
				created: datetime, modified: datetime, image_length_px: int, image_width_px: int,
				uuid: UUID):
		self.image_id = image_id
		self.herd_unit_id = herd_unit_id
		self.survey_id = survey_id
		self.img_key = img_key
		self.name = name
		self.in_training = in_training
		self.crops_generated = crops_generated
		self.opened_by_user_id = opened_by_user_id
		self.created = created
		self.modified = modified
		self.image_length_px = image_length_px
		self.image_width_px = image_width_px
		self.uuid = uuid
		self.image = None

	def setImage(self, imageData: np.ndarray):
		if isinstance(imageData, bytes):
			try:
				pilImage = PillowImage.open(io.BytesIO(imageData))
				pilImage = pilImage.convert('RGB')
				self.image = cv2.cvtColor(np.array(pilImage), cv2.COLOR_RGB2BGR)
			except Exception as e:
				print(f"Error converting bytes to image: {e}")
				self.image = None
		elif isinstance(imageData, np.ndarray):
			self.image = imageData
		else:
			print(f"Unsupported type: {type(imageData)}")
			self.image = None

	def getImage(self) -> Optional[np.ndarray]:
		if self.image is not None:
			return self.image

	def deleteImage(self):
		del self.image

	def serve(self, img_format: str):
		if self.image is not None:
			_, self.img_encoded = cv2.imencode(img_format, self.getImage(), [cv2.IMWRITE_JPEG_QUALITY, 100]) #type: ignore
		else:
			raise Exception(f'{self.name} has no image data')
		return self.img_encoded.tobytes()
		
	def serialize(self) -> dict:
		return {
			'image_id': self.image_id,
			'herd_unit_id': self.herd_unit_id,
			'survey_id': self.survey_id,
			'img_key': self.img_key,
			'name': self.name,
			'in_training': self.in_training,
			'crops_generated': self.crops_generated,
			'created': self.created,
			'modified': self.modified,
			'image_length_px': self.image_length_px,
			'image_width_px': self.image_width_px,
			'uuid': self.uuid
		}

class Prediction(CgOBJ):
	def __init__(self, pred_id: int, image_id: int, model_id: int, label: int, score: float, box_tx: int, box_ty: int,
				box_bx: int, box_by: int, created: datetime | None, reviewed_by_user_id: int, uuid: UUID):
		self.pred_id = pred_id
		self.image_id = image_id
		self.model_id = model_id
		self.label = label
		self.score = score
		self.dimensions = Box((box_tx, box_ty), (box_bx, box_by))
		self.reviewed_by_user_id = reviewed_by_user_id
		self.created = created 
		self.uuid = uuid
	
	def serialize(self) -> dict:
		return {
			'pred_id': self.pred_id,
			'image_id': self.image_id,
			'model_id': self.model_id,
			'dimensions': self.dimensions.serialize(),
			'score': self.score,
			'label': self.label,
			'created': self.created,
			'reviewed_by_user_id': self.reviewed_by_user_id,
			'uuid': self.uuid
		}
	
class Annotation(CgOBJ):
	def __init__(self, label_id: int, image_id: int, herd_unit_id: int,
				box_tx: int, box_ty: int, box_bx: int, box_by: int, annotation_id: int | None =None, created_by_user_id: int | None =None, 
				created: datetime | None =None, modified: datetime | None =None, uuid: UUID | None =None, pred_id: int | None=None):
		self.annotation_id = annotation_id
		self.label_id = label_id
		self.pred_id = pred_id
		self.image_id = image_id
		self.herd_unit_id = herd_unit_id
		self.dimensions = Box((box_tx, box_ty), (box_bx, box_by))
		self.created_by_user_id = created_by_user_id
		self.created = created
		self.modified = modified
		self.uuid = uuid
	
	def serialize(self) -> dict:
		return {
			'annotation_id': self.annotation_id,
			'label_id': self.label_id,
			'image_id': self.image_id,
			'herd_unit_id': self.herd_unit_id,
			'dimensions': self.dimensions.serialize(),
			'created': self.created,
			'modified': self.modified,
			'uuid': self.uuid
		}


class ReviewedArea(Image):
	def __init__(self,  image_id: int, name: str,  area_tx: int, area_ty: int, area_bx: int, 
				area_by: int, reviewed_area_length_px: int, reviewed_area_width_px: int,
				reviewed_area_id: int | None = None, created: datetime | None = None , 
				modified: datetime | None = None, reviewed_by_user_id: int | None = None, 
				uuid: UUID | None = None, ra_key: str | None = None):
		self.reviewed_area_id = reviewed_area_id
		self.image_id = image_id
		self.name = name
		self.dimensions = Box((area_tx, area_ty), (area_bx, area_by))
		self.created = created
		self.modified = modified
		self.reviewed_area_length_px = abs(area_ty - area_by)
		self.reviewed_area_width_px = abs(area_tx - area_bx)
		self.reviewed_by_user_id = reviewed_by_user_id
		self.uuid = uuid
		self.ra_key = ra_key

	def serialize(self) -> dict:
		return {
			'reviewed_area_id': self.reviewed_area_id,
			'image_id': self.image_id,
			'name': self.name,
			'dimensions': self.dimensions.serialize(),
			'created': self.created,
			'modified': self.modified,
			'reviewed_area_length_px': self.reviewed_area_length_px,
			'reviewed_area_width_px': self.reviewed_area_width_px,
			'ra_key': self.ra_key,
			'uuid': self.uuid,
		}
	
class PredictionCrop(Image):
	def __init__(self, image_id: int, pred_id: int, name: str, score: float, label: int, dimensions: Box, 
				boundingBox: Box, uuid: UUID, url: Optional[str]=None):
		self.image_id = image_id
		self.pred_id = pred_id
		self.name = name
		self.score = score
		self.label = label 
		self.dimensions = dimensions
		self.boundingBox = boundingBox
		self.approved = False
		self.uuid = uuid
	
	def serialize(self) -> dict:
		return {
			'image_id': self.image_id,
			'pred_id': self.pred_id,
			'name': self.name,
			'score': self.score,
			'label': self.label,
			'dimensions': self.dimensions.serialize(),
			'boundingBox': self.boundingBox.serialize(),
			'approved': self.approved,
			'uuid': self.uuid,
		}
#---------------------------------------------------------------------------------------------------------------------------#

class CropgenJSONPRovider(JSONProvider):
	def default(self, obj):
		if isinstance(obj, CgOBJ):
			return obj.serialize()
		raise TypeError(f"Object of type {obj.__class__.__name__} is not JSON serializable")


#---------------------------------------------------------------------------------------------------------------------------#

File: generator_web_app/src/types/generatorobjects.ts
// Typescript Class definition analogues for crop_generator objects 
// Author: Michael B. Lance
// Created: April 9, 2025
// Updated: November, 2025
//---------------------------------------------------------------------------------------------------------------------------//

import { mean } from 'lodash';
import type { Ref } from 'vue';

//---------------------------------------------------------------------------------------------------------------------------//s

export interface konvaBoxConf {
    width: number;
    height: number;
    x: number;
    y: number;
    scaleX: number,
    scaleY: number,
    rotation: number;
    stroke: string;
}

export interface tempRect {
    startPointX: number,
    startPointY: number,
    width: number,
    height: number
}

export interface coord {
    x: number,
    y: number
}

export interface BoxIntf {
    top_left: coord;
    bottom_right: coord;
}

export class Box implements BoxIntf {
    top_left: coord;
    bottom_right: coord;
    constructor(box: BoxIntf) {
        this.top_left = box.top_left;
        this.bottom_right = box.bottom_right;
    }
    getCenter(): number[] {
        let x = mean([this.top_left.x, this.bottom_right.x]);
        let y = mean([this.top_left.y, this.bottom_right.y]);
        return [x, y];
    }
    getPoints(): number[] {
        return [this.top_left.x, this.top_left.y, this.bottom_right.x, this.bottom_right.y];
    }
    getWidth(): number {
        return Math.abs(this.top_left.x - this.bottom_right.x);
    }
    getHeight(): number {
        return Math.abs(this.top_left.y - this.bottom_right.y);
    }
    to_dict() {
        return [this.top_left.x, this.top_left.y, this.bottom_right.x, this.bottom_right.y];

    }

}

//---------------------------------------------------------------------------------------------------------------------------//

export interface HerdUnitIntf {
    herd_unit_id: number;
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class HerdUnit implements HerdUnitIntf {
    herd_unit_id: number;
    name: string;
    created: Date;
    modified: Date;
    uuid: string;


    constructor(herdunit: HerdUnitIntf) {
        this.herd_unit_id = herdunit.herd_unit_id;
        this.name = herdunit.name;
        this.created = new Date(herdunit.created);
        this.modified = new Date(herdunit.modified);
        this.uuid = herdunit.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export type CGObject = {
    name: string;
    uuid: string;
    created: Date;
    modified: Date;
}

export interface ImageIntf {
    image_id: number;
    herd_unit_id: number;
    survey_id: number;
    img_key: string;
    name: string;
    in_training: boolean;
    crops_generated: number;
    created: Date;
    modified: Date;
    image_length_px: number;
    image_width_px: number;
    uuid: string;
}

export class Image implements ImageIntf {
    image_id: number;
    herd_unit_id: number;
    survey_id: number;
    img_key: string;
    name: string;
    in_training: boolean;
    crops_generated: number;
    created: Date;
    modified: Date;
    image_length_px: number;
    image_width_px: number;
    uuid: string;

    constructor(img: ImageIntf) {
        this.image_id = img.image_id;
        this.herd_unit_id = img.herd_unit_id;
        this.survey_id = img.survey_id;
        this.img_key = img.img_key;
        this.name = img.name;
        this.in_training = img.in_training;
        this.crops_generated = img.crops_generated;
        this.created = new Date(img.created);
        this.modified = new Date(img.modified);
        this.image_length_px = img.image_length_px;
        this.image_width_px = img.image_width_px;
        this.uuid = img.uuid
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface PredictionIntf {
    pred_id: number;
    image_id: number;
    model_id: number;
    dimensions: Box;
    score: number;
    label: number;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Prediction implements PredictionIntf {
    pred_id: number;
    image_id: number;
    model_id: number;
    dimensions: Box;
    score: number;
    label: number;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(pred: PredictionIntf) {
        this.pred_id = pred.pred_id;
        this.image_id = pred.image_id;
        this.model_id = pred.model_id;
        this.dimensions = new Box(pred.dimensions);
        this.score = pred.score;
        this.label = pred.label;
        this.created = new Date(pred.created);
        this.modified = new Date(pred.modified);
        this.uuid = pred.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface AnnotationIntf {
    annotation_id: number;
    label_id: number;
    image_id: number;
    herd_unit_id: number;
    dimensions: Box;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Annotation implements AnnotationIntf {
    annotation_id: number;
    label_id: number;
    image_id: number;
    herd_unit_id: number;
    dimensions: Box;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(annotation: AnnotationIntf) {
        this.annotation_id = annotation.annotation_id;
        this.label_id = annotation.label_id;
        this.image_id = annotation.image_id;
        this.herd_unit_id = annotation.herd_unit_id;
        this.dimensions = new Box(annotation.dimensions);
        this.created = new Date(annotation.created);
        this.modified = new Date(annotation.modified);
        this.uuid = annotation.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface ReviewedAreaIntf {
    reviewed_area_id: number;
    image_id: number;
    ra_key: string;
    name: string;
    dimensions: Box;
    created: Date;
    modified: Date;
    reviewed_area_length_px: number;
    reviewed_area_width_px: number;
    uuid: string;
    url: string;
}

export class ReviewedArea implements ReviewedAreaIntf {
    reviewed_area_id: number;
    image_id: number;
    ra_key: string;
    name: string;
    dimensions: Box;
    created: Date;
    modified: Date;
    reviewed_area_length_px: number;
    reviewed_area_width_px: number;
    uuid: string;
    url: string;

    constructor(crop: ReviewedAreaIntf) {

        this.reviewed_area_id = crop.reviewed_area_id;
        this.ra_key = crop.ra_key;
        this.image_id = crop.image_id;
        this.name = crop.name;
        this.dimensions = new Box(crop.dimensions);
        this.created = new Date(crop.created);
        this.modified = new Date(crop.modified);
        this.reviewed_area_length_px = crop.reviewed_area_length_px;
        this.reviewed_area_width_px = crop.reviewed_area_width_px;
        this.uuid = crop.uuid;
        this.url = crop.url;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface PredictionCropIntf {
    image_id: number;
    pred_id: number;
    name: string;
    score: number;
    label: number;
    dimensions: Box;
    boundingBox: Box;
    approved: boolean;
    uuid: string;
}

export class PredictionCrop implements PredictionCropIntf {
    image_id: number;
    pred_id: number;
    name: string;
    score: number;
    label: number;
    approved: boolean;
    dimensions: Box;
    boundingBox: Box;
    url: string;
    uuid: string;
    drawBox: boolean = false;

    constructor(predcrop: PredictionCropIntf, url: string) {
        this.image_id = predcrop.image_id;
        this.pred_id = predcrop.pred_id;
        this.name = predcrop.name;
        this.score = predcrop.score;
        this.label = predcrop.label;
        this.dimensions = new Box(predcrop.dimensions);
        this.boundingBox = new Box(predcrop.boundingBox);
        this.approved = predcrop.approved;
        this.uuid = predcrop.uuid;
        this.url = url;
    }
}
//---------------------------------------------------------------------------------------------------------------------------//

export interface RoleIntf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Role implements RoleIntf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(role: RoleIntf) {
        this.name = role.name;
        this.created = new Date(role.created);
        this.modified = new Date(role.modified);
        this.uuid = role.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface UserIntf {
    username: string;
    status: string;
    created: Date;
    modified: Date;
    last_login: Date;
    locale: string;
    uuid: string;
    roles: RoleIntf[];
}

export class User {
    username: string;
    status: string;
    created: Date;
    modified: Date;
    last_login: Date;
    locale: string;
    roles: Role[];
    uuid: string;

    constructor(usr: UserIntf) {
        this.username = usr.username;
        this.status = usr.status;
        this.created = new Date(usr.created);
        this.modified = new Date(usr.modified);
        this.last_login = new Date(usr.last_login);
        this.locale = usr.locale;
        this.roles = [];
        if (usr.roles != undefined) {
            for (const role of usr.roles) this.roles.push(new Role(role));
        }
        this.uuid = usr.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface OrganizationIntf {
    name: string;
    created: Date;
    modified: Date;
    logo_url: string | undefined;
    uuid: string;
}

export class Organization implements OrganizationIntf {
    name: string;
    created: Date;
    modified: Date;
    logo_url: string | undefined;
    uuid: string;

    constructor(Org: OrganizationIntf) {
        this.name = Org.name;
        this.created = new Date(Org.created);
        this.modified = new Date(Org.modified);
        this.logo_url = Org.logo_url;
        this.uuid = Org.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface ProjectIntf {
    project_id: number;
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Project implements ProjectIntf {
    project_id: number;
    name: string;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(Proj: ProjectIntf) {
        this.project_id = Proj.project_id;
        this.name = Proj.name;
        this.created = new Date(Proj.created);
        this.modified = new Date(Proj.modified);
        this.uuid = Proj.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface SchemaIntf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Schema implements SchemaIntf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(Schem: SchemaIntf) {
        this.name = Schem.name;
        this.created = new Date(Schem.created);
        this.modified = new Date(Schem.modified);
        this.uuid = Schem.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface LabelIntf {
    label_id: number;
    label: number;
    name: string;
    image_link: string;
    color: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Label implements LabelIntf {
    label_id: number;
    label: number;
    name: string;
    image_link: string;
    color: string;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(lbl: LabelIntf) {
        this.label_id = lbl.label_id;
        this.label = lbl.label;
        this.name = lbl.name;
        this.image_link = lbl.image_link;
        this.color = lbl.color;
        this.created = new Date(lbl.created);
        this.modified = new Date(lbl.modified);
        this.uuid = lbl.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface SurveyIntf {
    survey_id: number
    survey_date: Date;
    name: string;
    additional_info: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Survey implements SurveyIntf {
    survey_id: number;
    survey_date: Date;
    name: string;
    additional_info: string;
    created: Date;
    modified: Date;
    uuid: string;

    constructor(srvy: SurveyIntf) {
        this.survey_id = srvy.survey_id;
        this.survey_date = new Date(srvy.survey_date);
        this.name = srvy.name;
        this.additional_info = srvy.additional_info;
        this.created = new Date(srvy.created);
        this.modified = new Date(srvy.modified);
        this.uuid = srvy.uuid;
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface ModelIntf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
}

export class Model implements ModelIntf {
    name: string;
    created: Date;
    modified: Date;
    uuid: string;
    constructor(mdl: ModelIntf) {
        this.name = mdl.name;
        this.created = new Date(mdl.created);
        this.modified = new Date(mdl.modified);
        this.uuid = mdl.uuid
    }
}

//---------------------------------------------------------------------------------------------------------------------------//

export interface autoCropperBatch {
    images: Image[],
    predictions: Prediction[][],
    predictionCrops: PredictionCrop[][],
}

export interface cropVerifierBatch {
    crops: {
        [crop_uuid: string]: {
            'crop': ReviewedArea,
            'image'?: readonly [
                Ref<HTMLImageElement | null, HTMLImageElement | null>,
                Ref<"loading" | "error" | "loaded", "loading" | "error" | "loaded">
            ];
            'annotations': {
                [annotation_uuid: string]: Annotation
            },
            'annotationBoxes': {
                [annotation_uuid: string]: konvaBoxConf
            }
        }
    },

}

//---------------------------------------------------------------------------------------------------------------------------//

export default {};